internal/sysfs/select_windows.go

```diff
@@ -37,69 +33,181 @@ func syscall_select(n int, r, w, e *platform.FdSet, timeout *time.Duration) (int
 		time.Sleep(*timeout)
 		return 0, nil
 	}
-	if r.IsSet(wasiFdStdin) {
-		fileType, err := syscall.GetFileType(syscall.Stdin)
-		if err != nil {
-			return 0, err
-		}
-		if fileType&syscall.FILE_TYPE_CHAR != 0 {
-			res, err := pollNamedPipe(context.TODO(), syscall.Stdin, timeout)
-			if err != nil {
-				return -1, err
-			}
-			if !res {
-				r.Zero()
-				return 0, nil
-			}
-		}
-		r.Zero()
-		r.Set(wasiFdStdin)
-		return 1, nil
+
+	n, errno := selectAllHandles(context.TODO(), r, w, e, timeout)
+	if errno == 0 {
+		return n, nil
 	}
-	return -1, sys.ENOSYS
+	return n, errno
 }
 
-// pollNamedPipe polls the given named pipe handle for the given duration.
+// selectAllHandles emulates a general-purpose POSIX select on Windows.
 //
 // The implementation actually polls every 100 milliseconds until it reaches the given duration.
 // The duration may be nil, in which case it will wait undefinely. The given ctx is
-// used to allow for cancellation. Currently used only in tests.
-func pollNamedPipe(ctx context.Context, pipeHandle syscall.Handle, duration *time.Duration) (bool, error) {
+// used to allow for cancellation, and it is currently used only in tests.
+//
+// As indicated in the man page for select [1], r, w, e are modified upon completion:
+//
+// "Upon successful completion, the pselect() or select() function shall modify the objects pointed to by the readfds,
+// writefds, and errorfds arguments to indicate which file descriptors are ready for reading, ready for writing,
+// or have an error condition pending, respectively, and shall return the total number of ready descriptors in all the output sets"
+//
+// However, for our purposes, this may be pedantic because currently we do not check the values of r, w, e
+// after the invocation of select; thus, this behavior may be subject to change in the future for the sake of simplicity.
+//
+// [1]: https://linux.die.net/man/3/select
+func selectAllHandles(ctx context.Context, r, w, e *platform.FdSet, duration *time.Duration) (int, sys.Errno) {
+	nregular := r.Regular().Count() + w.Regular().Count() + e.Regular().Count()
+
+	nsocks := 0
+
+	rp, errno := peekAllPipes(r.Pipes())
+	npipes := rp.Count()
+
+	if errno != 0 {
+		r.Zero()
+		w.Zero()
+		e.Zero()
+		r.SetPipes(*rp)
+		return nregular + npipes, errno
+	}
+
+	// winsock_select mutates the given references, so we create copies.
+	rs, ws, es := r.Sockets().Copy(), w.Sockets().Copy(), e.Sockets().Copy()
+
 	// Short circuit when the duration is zero.
 	if duration != nil && *duration == time.Duration(0) {
-		bytes, err := peekNamedPipe(pipeHandle)
-		return bytes > 0, err
+		nsocks, errno = winsock_select(rs, ws, es, duration)
+	} else {
+		// Ticker that emits at every pollInterval.
```

Comment:
I presume this wasn't extracted to a function because the state "rs, ws, es, duration" or whatever else is needed here, isn't easy to create (and unit test). If it isn't or isn't harder, I might suggest extracting this part and testing it directly vs indirectly via selectAllHandles.

Regardless, if you don't mind doing a coverage pass. It feels like `selectAllHandles` accepts a context param which would allow reaching some edge cases here (e.g. context already done), but not sure if it is or not.. coverage could tell for sure.

Finally, any area where we rely on goroutines can have some subtle issues and we should be careful. In general stateful concurrent code is the sort of last resort, and I understand this may be the only way out for windows (vs say not supported). Maybe use the hammer test (for the high level function) to make sure the windows calls used below don't have any one-at-a-time things that should be locked. That or say too much.. in any case please consider what may be needed about any concurrency issues here.

Reply:
I have reworked all this bit and did a coverage pass, adding a few more tests. With the `select(timeout:=0)` trick there is no longer a user-controlled goroutine, just channels with ticks and timeouts as before.